Activity Log: Add unavailability banner

dmsnell opened the pull request:
In certain circumstances Rewind is unavailable and yet there is
something the customer can do about it. In these cases we want to
provide them a hint at what that might be and an opportunity to
act on that.

In this patch we're creting a banner to display in the Activity Log
to show when a Jetpack site is disconnected and also when VaultPress
is active but also able to transfer to Rewind.

<img width="657" alt="screen shot 2018-01-31 at 4 25 13 pm" src="https://user-images.githubusercontent.com/5431237/35648555-19265662-06a4-11e8-833e-12ba8927c602.png">

<img width="653" alt="screen shot 2018-01-31 at 4 22 42 pm" src="https://user-images.githubusercontent.com/5431237/35648558-1a82fa10-06a4-11e8-8833-07e54801163d.png">

**Questions**
 - Although I wrote up the JSX here I have no idea what we should be saying. I would appreciate some recommendations for the banner titles and descriptions.
 - @MichaelArestad can you recommend appropriate icons for the banner?
 - Are these banners even appropriate? There's a sync-issue here with disconnected Jetpack sites because it appears like the backend isn't as aware of sites disconnecting and reconnecting as Calypso itself it. Should we instead have some kind of more-global notice of degraded service while disconnected and leave this banner out of here?
 - The VaultPress banner links to the credentials form and the disconnected notice links to the `wp-admin` of the Jetpack site. Is this right?

a8ci18n commented on client/my-sites/stats/activity-log/unavailability-notice.jsx:
```diff
@@ -0,0 +1,60 @@
+				<Banner
+					icon="history"
+					href={ adminUrl }
+					title={ translate( 'The site cannot be accessed' ) }
```
Hi! I've found a possible matching string that has already been translated **26** times:
`translate( 'This site cannot be accessed.' )` *ES Score: 14*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

MichaelArestad commented:
> The VaultPress banner links to the credentials form

Is this after they have enabled rewind and somehow enabled Vaultpress again? Or is this before they have switched?

> and the disconnected notice links to the wp-admin of the Jetpack site. Is this right?

What is the user supposed to do to fix this? Reconnect Jetpack? Or is it something else wrong with it like a communications error?

dmsnell commented:
> Is this after they have enabled rewind and somehow enabled Vaultpress again? Or is this before they have switched?

@MichaelArestad it would presumably only ever exist before since once they transfer they aren't on VaultPress again. I suppose if they "went back" it could appear again. It's meant to communicate "you would otherwise be able to Rewind but since you're on VaultPress we can't do it."

> What is the user supposed to do to fix this? Reconnect Jetpack? Or is it something else wrong with it like a communications error?

Reconnecting Jetpack is the solution but the problem is unknown. The problem is opaquely "we aren't connected to your site" and we don't know why. I suppose it's possible with intentional disconnects for us to know that. Otherwise the site could be down, a firewall could be in the way, or any other issue…

MichaelArestad commented:
> @MichaelArestad it would presumably only ever exist before since once they transfer they aren't on VaultPress again. I suppose if they "went back" it could appear again. It's meant to communicate "you would otherwise be able to Rewind but since you're on VaultPress we can't do it."

I thought we already had something similar already. Can you confirm @eliorivero ?

> Reconnecting Jetpack is the solution but the problem is unknown. The problem is opaquely "we aren't connected to your site" and we don't know why. I suppose it's possible with intentional disconnects for us to know that. Otherwise the site could be down, a firewall could be in the way, or any other issue…

Hmm. In that case, let's have them contact support to further help them identify a specific cause and solution.

**Icon**
`notice`

**Line 1**
> We are having trouble communicating with your site

**Line 2**
> We are unable to back up or rewind your site until communications are restored.

**Button**
> Get help

This might also be better as a Notice instead of a Banner.

dmsnell commented:
> This might also be better as a Notice instead of a Banner.

Since this could be a persistent condition I felt like a banner might be way too annoying. It's a state of the Rewind/Activity Log system but it's not something that just randomly pops up out of nowhere. It's unlikely to change within or across multiple sessions.

benhuberman commented:
I like @MichaelArestad's suggested copy direction. I'd slightly tweak to avoid (as much as possible) repeating similar words/constructions in the two sentences, though:
**Line 1**:
>  The connection to your site isn't working as it should.
**Line 2**:
> We can't back up or rewind your site until we reconnect to it.
**Button**:
> Get help

eliorivero commented on client/my-sites/stats/activity-log/index.jsx:
```diff
@@ -494,6 +495,7 @@ class ActivityLog extends Component {
 				<SidebarNavigation />
 				<StatsNavigation selectedItem={ 'activity' } siteId={ siteId } slug={ slug } />
 				{ siteId && <ActivityLogUpgradeNotice siteId={ siteId } /> }
+				{ 'unavailable' === rewindState.state && <UnavailabilityNotice siteId={ siteId } /> }
```
In the line before we make sure `siteId` is available before passing it as a prop. Maybe we should do the same here.
```js
{ ( siteId && 'unavailable' === rewindState.state ) && <UnavailabilityNotice siteId={ siteId } /> }
```

dmsnell replied:
hm. a good point. in this case there would be no bug I think because we couldn't get `unavailable` if there were no `siteId`, but it's probably just as good or better to include the extra check

eliorivero commented on client/my-sites/stats/activity-log/unavailability-notice.jsx:
```diff
@@ -0,0 +1,60 @@
+	};
+};
+
+export default connect( mapStateToProps )( localize( UnavailabilityNotice ) );
```
In most places we (even you) placed the contents of `mapStateToProps` inline in the connect without collecting it first in a constant. Why this pattern now?

dmsnell replied:
Whether I've inlined it I've always preferred explicitly separating them for consistency and explicitness. I've not made a big deal about it because we share a variety of preferences and it doesn't have a big impact. Also, inlining `mapDispatchToProps` when we have no arguments seems much more tolerable than `mapStateToProps` which _always_ needs args

a8ci18n commented on client/my-sites/stats/activity-log/unavailability-notice.jsx:
```diff
@@ -0,0 +1,60 @@
+				<Banner
+					icon="history"
+					href={ adminUrl }
+					title={ translate( 'The site is not connected' ) }
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'The site is not connected.' )` *ES Score: 12*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/rewind/show-unavailability-report)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

dmsnell commented:
@eliorivero I've added the suggested guard

@MichaelArestad and @benhuberman I updated the wording again in the disconnected notice to better reflect my current understanding of that condition. It doesn't appear to turn on flippantly at all. I've had a site whose web server has been off for a week and it's still not reporting `no_connected_jetpack` so I think the flag will activate when we manually disconnect vs. when the site is simply unreachable.

eliorivero commented:
I tried disconnecting Jetpack to test this working but upon doing it so, I got this screen

<img width="719" alt="captura de pantalla 2018-02-09 a la s 11 32 33" src="https://user-images.githubusercontent.com/1041600/36032506-015c828c-0d8d-11e8-9db9-269c738d33e8.png">

MichaelArestad commented:
@dmsnell the subtle change makes sense and is more accurate. 👍

dmsnell commented:
Good catch @eliorivero - I don't know what happened but I had `failureReason` in there instead of `reason` for no explainable reason :wink:

Fixed that and it's working now. Also, I tested by dispatching this event…

```js
dispatch( {
    type: 'REWIND_STATE_REQUEST',
    siteId: SOME_SITE_ID,
    meta: {
        dataLayer: {
            trackRequest: true,
            data: {
                state: 'unavailable',
                reason: 'no_connected_jetpack',
                last_updated: 1235466
            }
        },
    },
} )
```

I'm investigating the `no_connected_jetpack` because there were assumption about it I made that were wrong. I'm not sure if you manually disconnected or not but that's mainly what it's testing. it should either appear after manually disconnecting Jetpack or after about a month of not being able to reach the site.

thanks!
